File: pages/20260824170700_holding_topology_passing/analysis/ruby/passing_four_hand.rb
require_relative "passing_schedule"

module PassingAnalysis
  module FourHand
    HAND_CYCLE = [1, 3, 0, 2].freeze
    HAND_PERIOD = 4

    module_function

    def schedule(source, hold_fours: true)
      heights = parse(source)
      period = heights.length
      ball_count = heights.sum / period
      raise ArgumentError, "four-hand object count must be a positive integer" unless
        ball_count.is_a?(Integer) && ball_count.positive?

      highest = heights.max
      depth = [highest, 1].max
      landing = Array.new(4) { Array.new(depth) { [] } }
      intro = (0...ball_count).to_a
      cycle_tosses = []
      init_complete = false
      beat = 0
      start_key = nil
      cycle_length = 0

      4000.times do
        available = landing.map(&:shift)
        landing.each { |queue| queue << [] }
        from_hand = HAND_CYCLE.fetch(beat % HAND_PERIOD)
        recorded = toss(
          heights.fetch(beat % period), landing, available, intro, from_hand, beat, hold_fours, init_complete
        )
        raise ArgumentError, "prop landing with no toss at beat #{beat}" if available.any? { |queue| !queue.empty? }

        cycle_tosses.concat(recorded)
        if init_complete
          if start_key.nil?
            start_key = Marshal.dump(landing)
          elsif beat.positive? && (beat % period).zero? && (beat % HAND_PERIOD).zero? &&
              Marshal.dump(landing) == start_key
            cycle_tosses.pop while !cycle_tosses.empty? && cycle_tosses.last.beat == beat
            cycle_length = beat
            break
          end
        elsif intro.empty? && ((beat + 1) % period).zero? && ((beat + 1) % HAND_PERIOD).zero?
          init_complete = true
          beat = -1
        end
        beat += 1
      end
      raise ArgumentError, "four-hand pattern did not repeat" if cycle_length.zero?

      Schedule.new(
        pattern: nil,
        ball_count: ball_count,
        highest: highest,
        cycle_tosses: cycle_tosses,
        cycle_length: cycle_length,
        period: period,
        hand_count: 4,
        body_count: 2,
        hand_period: HAND_PERIOD
      )
    end

    def parse(source)
      text = source.to_s.strip
      raise ArgumentError, "empty four-hand siteswap" if text.empty?
      raise ArgumentError, "four-hand siteswap is a digit string, not JL passing" if text.include?("<")

      heights = text.chars.map { |character| throw_height(character) }
      raise ArgumentError, "four-hand pattern has no throws" if heights.empty?
      raise ArgumentError, "four-hand object count must be a positive integer" unless
        (heights.sum % heights.length).zero?

      heights
    end

    def throw_height(character)
      return Integer(character) if character =~ /[0-9]/
      return 10 + character.ord - 97 if character =~ /[a-z]/

      raise ArgumentError, "unsupported four-hand throw: #{character}"
    end

    def toss(height, landing, available, intro, from_hand, beat, hold_fours, record)
      from_body = from_hand / 2
      from_contact = from_hand % 2
      if height.zero?
        raise ArgumentError, "prop landing on 0 toss at beat #{beat}" unless available.fetch(from_hand).empty?
        return [] unless record

        return [Event.new(
          beat: beat, height: 0, ball: nil, from_body: from_body, from_contact: from_contact,
          from_hand: from_hand, to_body: from_body, to_contact: from_contact, to_hand: from_hand,
          hold: false, pass: false, socket_index: 0, kind: "empty"
        )]
      end

      ball = available.fetch(from_hand).shift || intro.shift
      raise ArgumentError, "no prop available at beat #{beat}" if ball.nil?

      to_hand = HAND_CYCLE.fetch((beat + height) % HAND_PERIOD)
      to_body = to_hand / 2
      to_contact = to_hand % 2
      hold = hold_fours && height == 4 && from_hand == to_hand
      landing.fetch(to_hand).fetch(height - 1) << ball
      return [] unless record

      [Event.new(
        beat: beat, height: height, ball: ball, from_body: from_body, from_contact: from_contact,
        from_hand: from_hand, to_body: to_body, to_contact: to_contact, to_hand: to_hand,
        hold: hold, pass: to_body != from_body, socket_index: 0, kind: hold ? "hold" : "throw"
      )]
    end
  end
end

File: pages/20260824170700_holding_topology_passing/analysis/julia/passing_four_hand.jl
const FOUR_HAND_CYCLE = [1, 3, 0, 2]
const FOUR_HAND_PERIOD = 4

function four_hand_throw_height(character)
    if '0' <= character <= '9'
        return Int(character - '0')
    end
    if 'a' <= character <= 'z'
        return 10 + Int(character - 'a')
    end
    throw(ArgumentError("unsupported four-hand throw: $character"))
end

function parse_four_hand(source)
    text = strip(string(source))
    isempty(text) && throw(ArgumentError("empty four-hand siteswap"))
    occursin("<", text) && throw(ArgumentError("four-hand siteswap is a digit string, not JL passing"))
    heights = [four_hand_throw_height(character) for character in text]
    isempty(heights) && throw(ArgumentError("four-hand pattern has no throws"))
    sum(heights) % length(heights) == 0 ||
        throw(ArgumentError("four-hand object count must be a positive integer"))
    return heights
end

function four_hand_toss!(height, landing, available, intro, from_hand, beat, hold_fours, record)
    from_body = from_hand ÷ 2
    from_contact = from_hand % 2
    if height == 0
        isempty(available[from_hand + 1]) ||
            throw(ArgumentError("prop landing on 0 toss at beat $beat"))
        record || return PassingEvent[]
        return [PassingEvent(
            beat, 0, nothing, from_body, from_contact, from_hand, from_body, from_contact, from_hand,
            false, false, 0, "empty",
        )]
    end
    ball = isempty(available[from_hand + 1]) ? (isempty(intro) ? nothing : popfirst!(intro)) :
        popfirst!(available[from_hand + 1])
    ball === nothing && throw(ArgumentError("no prop available at beat $beat"))
    to_hand = FOUR_HAND_CYCLE[((beat + height) % FOUR_HAND_PERIOD) + 1]
    to_body = to_hand ÷ 2
    to_contact = to_hand % 2
    hold = hold_fours && height == 4 && from_hand == to_hand
    push!(landing[to_hand + 1][height], ball)
    record || return PassingEvent[]
    return [PassingEvent(
        beat, height, ball, from_body, from_contact, from_hand, to_body, to_contact, to_hand,
        hold, to_body != from_body, 0, hold ? "hold" : "throw",
    )]
end

function schedule_four_hand(source; hold_fours=true)
    heights = parse_four_hand(source)
    period = length(heights)
    ball_count = sum(heights) ÷ period
    (ball_count isa Integer && ball_count >= 1) ||
        throw(ArgumentError("four-hand object count must be a positive integer"))
    highest = maximum(heights)
    depth = max(highest, 1)
    landing = [[Int[] for _ in 1:depth] for _ in 1:4]
    intro = collect(0:(ball_count - 1))
    cycle_tosses = PassingEvent[]
    init_complete = false
    beat = 0
    start_key = nothing
    cycle_length = 0
    for _ in 1:4000
        available = [isempty(queue) ? Int[] : popfirst!(queue) for queue in landing]
        for queue in landing
            push!(queue, Int[])
        end
        from_hand = FOUR_HAND_CYCLE[(beat % FOUR_HAND_PERIOD) + 1]
        recorded = four_hand_toss!(
            heights[(beat % period) + 1], landing, available, intro, from_hand, beat, hold_fours, init_complete,
        )
        any(!isempty, available) && throw(ArgumentError("prop landing with no toss at beat $beat"))
        append!(cycle_tosses, recorded)
        if init_complete
            if start_key === nothing
                start_key = landing_key(landing)
            elseif beat > 0 && beat % period == 0 && beat % FOUR_HAND_PERIOD == 0 &&
                    landing_key(landing) == start_key
                while !isempty(cycle_tosses) && last(cycle_tosses).beat == beat
                    pop!(cycle_tosses)
                end
                cycle_length = beat
                break
            end
        elseif isempty(intro) && (beat + 1) % period == 0 && (beat + 1) % FOUR_HAND_PERIOD == 0
            init_complete = true
            beat = -1
        end
        beat += 1
    end
    cycle_length == 0 && throw(ArgumentError("four-hand pattern did not repeat"))
    return PassingSchedule(
        nothing, ball_count, highest, cycle_tosses, cycle_length, period, 4, 2, FOUR_HAND_PERIOD,
    )
end
